src/core/lib/surface/call.c

```diff
@@ -984,6 +984,10 @@ static void continue_receiving_slices(grpc_exec_ctx *exec_ctx,
       call->receiving_message = 0;
       grpc_byte_stream_destroy(exec_ctx, call->receiving_stream);
       call->receiving_stream = NULL;
+      if (call->status[STATUS_FROM_API_OVERRIDE].code != GRPC_STATUS_OK) {
```

Comment:
(on top of the current fix not being thread-safe... but otherwise it solves the leak).